src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,382 @@
+    float distance = (float)(fabs(a*x+b*y+c))/sqrt(a*a+b*b);
+    return (int)distance;
+}
+
+void initialize_line(Line *L, const flex_dict &p1, const flex_dict &p2) {
+    Point P1; make_point(&P1, p1);
+    Point P2; make_point(&P2, p2);
+    float start_x = P1.x;
+    float start_y = P1.y;
+    float end_x = P2.x;
+    float end_y = P2.y;
+    if (start_x == start_y) {
+        L->slope = INT_MAX;
+    } else {
+        L->slope = ((float)(end_y - start_y))/((float)(end_x - start_x));
+    }
+    L->a = L->slope;
+    L->b = -1;
+    L->c = start_y - L->slope * start_x;
+}
+
+template<typename T>
+std::vector<T> slice(std::vector<T> const &v, int m, int n) {
+    auto first = v.cbegin() + m;
+    auto last = v.cbegin() + n + 1;
+    std::vector<T> vec(first, last);
+    return vec;
+}
+
+flex_list ramer_douglas_peucker(flex_list stroke, float epsilon) {
+    float dmax = 0;
+    int index = 0;
+    flex_list compressed_stroke;
+    int num_points = stroke.size();
+    Line L;
+    const flex_dict &first_point = stroke[0].get<flex_dict>();
+    const flex_dict &last_point = stroke[num_points-1].get<flex_dict>();
+    initialize_line(&L, first_point, last_point);
+    for (int i = 0; i < num_points; i++) {
+        float d = line_to_point_distance(&L, stroke[i].get<flex_dict>());
+        if (d > dmax) {
+            index = i;
+            dmax = d;
+        }
+    }
+    if (dmax > epsilon) {
+        flex_list first_slice = slice(stroke, 0, index-1);
+        flex_list last_slice = slice(stroke, index, num_points-1);
+        compressed_stroke = ramer_douglas_peucker(first_slice, epsilon);
+        flex_list rec_results_2 = ramer_douglas_peucker(last_slice, epsilon);
+        compressed_stroke.insert(compressed_stroke.end(), 
+            rec_results_2.begin(), rec_results_2.end());
+    } else {
+        compressed_stroke.push_back(first_point);
+        compressed_stroke.push_back(last_point);
+    }
+    return compressed_stroke;
+}
+
+flex_list simplify_drawing(flex_list raw_drawing) {
+    size_t num_strokes = raw_drawing.size();
+    size_t min_x = INT_MAX;
+    size_t max_x = 0;
+    size_t min_y = INT_MAX;
+    size_t max_y = 0;
+    flex_list simplified_drawing;
+    Point P;
+    // Compute bounds of the drawing
+    for (size_t i = 0; i < num_strokes; i++) {
+        const flex_list &stroke = raw_drawing[i].get<flex_list>();
+        for (size_t j = 0; j < stroke.size(); j++) {
+            const flex_dict &point = stroke[j].get<flex_dict>();
+            make_point(&P, point);
+            float x = P.x;
+            float y = P.y;
+            if (x < min_x) min_x = x;
+            if (x > max_x) max_x = x;
+            if (y < min_y) min_y = y;
+            if (y > max_y) max_y = y;
+        }
+    }
+    // Align the drawing to top-left corner and scale to [0,255]
+    for (size_t i = 0; i < num_strokes; i++) {
+        flex_list &stroke = raw_drawing[i].mutable_get<flex_list>();
```

Comment:
You're using `mutable_get` but not actually mutating the result. In this case, since `raw_drawing` was passed by value, you could actually just mutate the result. This would save on a number of allocations.